KEYVALUE TYPE="" incorrectly attempted to be parsed as numeric

The `tupleparse.parse_keyvalue()` function handles a missing TYPE attribute on the KEYVALUE element correctly by falling back to use the VALUETYPE attribute. However, it does not handle a TYPE attribute value that is the empty string. In that case, it interprets the key value as a numeric value.

The end2end tests on the Brocade server in the SMI-S testlab fails due to that:

```
[1m============================= test session starts ==============================[0m
platform linux -- Python 3.4.3, pytest-3.8.2, py-1.5.3, pluggy-0.7.1
rootdir: /home/kschopmeyer/pywbem/githubpywbem/pywbem, inifile:
plugins: cov-2.5.1
collected 103 items

test_profile_generic.py s.ss.s.ss..ssssssssssssss.........s..........ssssssssssssssssssssssssssss.ssss
test_profile_snia_server.py F
test_server.py ........................

=================================== FAILURES ===================================
[1m[31m______ Test_SNIA_Server_Profile.test_instances[server_definition=Brocade] ______[0m

self = <pywbem.tupleparse.TupleParser object at 0x2b3213646518>
data = 'WBEMSolutions_ObjectManager', cimtype = ''

[1m    def unpack_numeric(self, data, cimtype):[0m
[1m        """[0m
[1m            Unpack a string value of a numeric CIM type and return its CIM data[0m
[1m            type object, or None.[0m
[1m    [0m
[1m            data (unicode string): CIM-XML string value, or None (in which case[0m
[1m              None is returned).[0m
[1m    [0m
[1m            cimtype (string): CIM data type name (e.g. 'uint8'), or None (in which[0m
[1m              case the value is returned as a Python int/long or float).[0m
[1m            """[0m
[1m    [0m
[1m        if data is None:[0m
[1m            return None[0m
[1m    [0m
[1m        # DSP0201 defines numeric values to be whitespace-tolerant[0m
[1m        data = data.strip()[0m
[1m    [0m
[1m        # Decode the CIM-XML string representation into a Python number[0m
[1m        #[0m
[1m        # Some notes:[0m
[1m        # * For integer numbers, only decimal and hexadecimal strings are[0m
[1m        #   allowed - no binary or octal.[0m
[1m        # * In Python 2, int() automatically returns a long, if needed.[0m
[1m        # * For real values, DSP0201 defines a subset of the syntax supported[0m
[1m        #   by Python float(), including the special states Inf, -Inf, NaN. The[0m
[1m        #   only known difference is that DSP0201 requires a digit after the[0m
[1m        #   decimal dot, while Python does not.[0m
[1m        if CIMXML_HEX_PATTERN.match(data):[0m
[1m            value = int(data, 16)[0m
[1m        else:[0m
[1m            try:[0m
[1m>               value = int(data)[0m
[1m[31mE               ValueError: invalid literal for int() with base 10: 'WBEMSolutions_ObjectManager'[0m

[1m[31m../../pywbem/tupleparse.py[0m:2284: ValueError

[33mDuring handling of the above exception, another exception occurred:[0m

self = <pywbem.tupleparse.TupleParser object at 0x2b3213646518>
data = 'WBEMSolutions_ObjectManager', cimtype = ''

[1m    def unpack_numeric(self, data, cimtype):[0m
[1m        """[0m
[1m            Unpack a string value of a numeric CIM type and return its CIM data[0m
[1m            type object, or None.[0m
[1m    [0m
[1m            data (unicode string): CIM-XML string value, or None (in which case[0m
[1m              None is returned).[0m
[1m    [0m
[1m            cimtype (string): CIM data type name (e.g. 'uint8'), or None (in which[0m
[1m              case the value is returned as a Python int/long or float).[0m
[1m            """[0m
[1m    [0m
[1m        if data is None:[0m
[1m            return None[0m
[1m    [0m
[1m        # DSP0201 defines numeric values to be whitespace-tolerant[0m
[1m        data = data.strip()[0m
[1m    [0m
[1m        # Decode the CIM-XML string representation into a Python number[0m
[1m        #[0m
[1m        # Some notes:[0m
[1m        # * For integer numbers, only decimal and hexadecimal strings are[0m
[1m        #   allowed - no binary or octal.[0m
[1m        # * In Python 2, int() automatically returns a long, if needed.[0m
[1m        # * For real values, DSP0201 defines a subset of the syntax supported[0m
[1m        #   by Python float(), including the special states Inf, -Inf, NaN. The[0m
[1m        #   only known difference is that DSP0201 requires a digit after the[0m
[1m        #   decimal dot, while Python does not.[0m
[1m        if CIMXML_HEX_PATTERN.match(data):[0m
[1m            value = int(data, 16)[0m
[1m        else:[0m
[1m            try:[0m
[1m                value = int(data)[0m
[1m            except ValueError:[0m
[1m                try:[0m
[1m>                   value = float(data)[0m
[1m[31mE                   ValueError: could not convert string to float: 'WBEMSolutions_ObjectManager'[0m

[1m[31m../../pywbem/tupleparse.py[0m:2287: ValueError

[33mDuring handling of the above exception, another exception occurred:[0m

self = <test_profile_snia_server.Test_SNIA_Server_Profile object at 0x2b32120642b0>
wbem_connection = WBEMConnection(url='https://10.1.134.163', creds=('Administrator', ...), conn_id='2-4347', default_namespace='root/cim...ck=None, ca_certs=None, no_verification=True, timeout=10, use_pull_operations=False, stats_enabled=False, recorders=[])

[1m    def test_instances(self, wbem_connection):  # noqa: F811[0m
[1m        """[0m
[1m            Test the instances of the profile.[0m
[1m            """[0m
[1m        self.init_profile(wbem_connection)[0m
[1m    [0m
[1m        central_inst_paths = self.server.get_central_instances([0m
[1m            self.profile_inst.path,[0m
[1m            central_class=self.profile_definition['central_class'],[0m
[1m            scoping_class=self.profile_definition['scoping_class'],[0m
[1m            scoping_path=self.profile_definition['scoping_path'],[0m
[1m            reference_direction=self.profile_definition['reference_direction'])[0m
[1m    [0m
[1m        # Check that there is one central instance[0m
[1m        assert len(central_inst_paths) == 1[0m
[1m        central_inst_path = central_inst_paths[0][0m
[1m    [0m
[1m        # Test the CIM_ObjectManager central instance[0m
[1m        self.assert_instance_of(central_inst_path, 'CIM_ObjectManager')[0m
[1m        central_inst = self.conn.GetInstance(central_inst_path)[0m
[1m        self.assert_instance_of(central_inst, 'CIM_ObjectManager')[0m
[1m        self.assert_instance_consistency(central_inst, central_inst_path)[0m
[1m        mandatory_property_list = [[0m
[1m            'SystemCreationClassName', 'SystemName',[0m
[1m            'CreationClassName', 'Name',[0m
[1m            'ElementName', 'Description',[0m
[1m            'OperationalStatus', 'Started',[0m
[1m        ][0m
[1m        self.assert_mandatory_properties(central_inst, mandatory_property_list)[0m
[1m    [0m
[1m        # Test the associated CIM_Namespace instances[0m
[1m        far_insts, _ = self.assert_association([0m
[1m            source_path=central_inst_path,[0m
[1m            source_role='Antecedent',[0m
[1m            assoc_class='CIM_NamespaceInManager',[0m
[1m            far_role='Dependent',[0m
[1m>           far_class='CIM_Namespace')[0m

[1m[31mtest_profile_snia_server.py[0m:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[1m[31mpytest_end2end.py[0m:507: in assert_association
[1m    Role=source_role)[0m
[1m[31m../../pywbem/cim_operations.py[0m:3688: in References
[1m    **extra)[0m
[1m[31m../../pywbem/cim_operations.py[0m:1708: in _imethodcall
[1m    tup_tree = tp.parse_cim(tt_)[0m
[1m[31m../../pywbem/tupleparse.py[0m:400: in parse_cim
[1m    child = self.one_child(tup_tree, ['MESSAGE', 'DECLARATION'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1624: in parse_message
[1m    'SIMPLEEXPRSP', 'MULTIEXPRSP'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1879: in parse_simplersp
[1m    child = self.one_child(tup_tree, ['METHODRESPONSE', 'IMETHODRESPONSE'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1938: in parse_imethodresponse
[1m    ['ERROR', 'IRETURNVALUE', 'PARAMVALUE']))[0m
[1m[31m../../pywbem/tupleparse.py[0m:315: in list_of_various
[1m    result.append(self.parse_any(child))[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:2024: in parse_ireturnvalue
[1m    'VALUE.INSTANCEWITHPATH'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:363: in list_of_same
[1m    result.append(self.parse_any(child))[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:699: in parse_value_objectwithpath
[1m    inst_path = self.parse_instancepath(k[0])[0m
[1m[31m../../pywbem/tupleparse.py[0m:904: in parse_instancepath
[1m    inst_path = self.parse_instancename(k[1])[0m
[1m[31m../../pywbem/tupleparse.py[0m:979: in parse_instancename
[1m    for key_bind in self.list_of_various(tup_tree, ['KEYBINDING']):[0m
[1m[31m../../pywbem/tupleparse.py[0m:315: in list_of_various
[1m    result.append(self.parse_any(child))[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1024: in parse_keybinding
[1m    child = self.one_child(tup_tree, ['KEYVALUE', 'VALUE.REFERENCE'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:499: in parse_value_reference
[1m    'INSTANCENAME'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:904: in parse_instancepath
[1m    inst_path = self.parse_instancename(k[1])[0m
[1m[31m../../pywbem/tupleparse.py[0m:979: in parse_instancename
[1m    for key_bind in self.list_of_various(tup_tree, ['KEYBINDING']):[0m
[1m[31m../../pywbem/tupleparse.py[0m:315: in list_of_various
[1m    result.append(self.parse_any(child))[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1024: in parse_keybinding
[1m    child = self.one_child(tup_tree, ['KEYVALUE', 'VALUE.REFERENCE'])[0m
[1m[31m../../pywbem/tupleparse.py[0m:278: in one_child
[1m    return self.parse_any(child)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2053: in parse_any
[1m    return func(tup_tree)  # a bound method, i.e. self is implicit[0m
[1m[31m../../pywbem/tupleparse.py[0m:1078: in parse_keyvalue
[1m    return self.unpack_single_value(data, cimtype)[0m
[1m[31m../../pywbem/tupleparse.py[0m:2169: in unpack_single_value
[1m    return self.unpack_numeric(data, cimtype)[0m
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pywbem.tupleparse.TupleParser object at 0x2b3213646518>
data = 'WBEMSolutions_ObjectManager', cimtype = ''

[1m    def unpack_numeric(self, data, cimtype):[0m
[1m        """[0m
[1m            Unpack a string value of a numeric CIM type and return its CIM data[0m
[1m            type object, or None.[0m
[1m    [0m
[1m            data (unicode string): CIM-XML string value, or None (in which case[0m
[1m              None is returned).[0m
[1m    [0m
[1m            cimtype (string): CIM data type name (e.g. 'uint8'), or None (in which[0m
[1m              case the value is returned as a Python int/long or float).[0m
[1m            """[0m
[1m    [0m
[1m        if data is None:[0m
[1m            return None[0m
[1m    [0m
[1m        # DSP0201 defines numeric values to be whitespace-tolerant[0m
[1m        data = data.strip()[0m
[1m    [0m
[1m        # Decode the CIM-XML string representation into a Python number[0m
[1m        #[0m
[1m        # Some notes:[0m
[1m        # * For integer numbers, only decimal and hexadecimal strings are[0m
[1m        #   allowed - no binary or octal.[0m
[1m        # * In Python 2, int() automatically returns a long, if needed.[0m
[1m        # * For real values, DSP0201 defines a subset of the syntax supported[0m
[1m        #   by Python float(), including the special states Inf, -Inf, NaN. The[0m
[1m        #   only known difference is that DSP0201 requires a digit after the[0m
[1m        #   decimal dot, while Python does not.[0m
[1m        if CIMXML_HEX_PATTERN.match(data):[0m
[1m            value = int(data, 16)[0m
[1m        else:[0m
[1m            try:[0m
[1m                value = int(data)[0m
[1m            except ValueError:[0m
[1m                try:[0m
[1m                    value = float(data)[0m
[1m                except ValueError:[0m
[1m                    raise ParseError([0m
[1m                        _format("Invalid numeric value {0!A}", data),[0m
[1m>                       conn_id=self.conn_id)[0m
[1m[31mE                   pywbem.exceptions.ParseError: Invalid numeric value 'WBEMSolutions_ObjectManager', Connection id: 2-4347[0m

[1m[31m../../pywbem/tupleparse.py[0m:2291: ParseError
```

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -519,6 +519,11 @@ This version contains all fixes up to pywbem 0.12.4.
   profiles, and that is used to drive profile discovery related tests against
   WBEM servers.
 
+* Added toleration support in the CIM-XML response parsing for WBEM servers
+  that return attribute `TYPE` with an empty string instead of omitting it.
+  As part of that, improved the checking for valid values of the TYPE
+  attribute. See issue #1564.
+
 **Known issues:**
 
 * See `list of open issues`_.

--- a/pywbem/tupleparse.py
+++ b/pywbem/tupleparse.py
@@ -95,6 +95,7 @@ __all__ = []
 
 
 CIMXML_HEX_PATTERN = re.compile(r'^(\+|\-)?0[xX][0-9a-fA-F]+$')
+NUMERIC_CIMTYPE_PATTERN = re.compile(r'^([su]int(8|16|32|64)|real(32|64))$')
 
 
 def filter_tuples(list_):
@@ -1058,12 +1059,20 @@ class TupleParser(object):
         data = self.pcdata(tup_tree)
         attrl = attrs(tup_tree)
 
-        valuetype = attrl.get('VALUETYPE', 'string')
+        valuetype = attrl.get('VALUETYPE', None)
         cimtype = attrl.get('TYPE', None)
+        cimtype_origin = "TYPE attribute ({0!r}) of {1} element, with " \
+            "fallback to VALUETYPE attribute ({2!r})". \
+            format(cimtype, name(tup_tree), valuetype)
+
+        # Tolerate that some WBEM servers return TYPE="" instead of omitting
+        # TYPE (e.g. the WBEM Solutions server).
+        if cimtype == '':
+            cimtype = None
 
         # Default the CIM type from VALUETYPE if not specified in TYPE
         if cimtype is None:
-            if valuetype == 'string':
+            if valuetype is None or valuetype == 'string':
                 cimtype = 'string'
             elif valuetype == 'boolean':
                 cimtype = 'boolean'
@@ -1075,7 +1084,7 @@ class TupleParser(object):
                             "value {1!A}", name(tup_tree), valuetype),
                     conn_id=self.conn_id)
 
-        return self.unpack_single_value(data, cimtype)
+        return self.unpack_single_value(data, cimtype, cimtype_origin)
 
     #
     # Object definition elements
@@ -2129,6 +2138,8 @@ class TupleParser(object):
         """
 
         valtype = attrs(tup_tree)['TYPE']
+        valtype_origin = "TYPE attribute ({0!r}) of {0} element". \
+            format(valtype, name(tup_tree))
 
         raw_val = self.list_of_matching(tup_tree, ['VALUE', 'VALUE.ARRAY'])
 
@@ -2144,12 +2155,12 @@ class TupleParser(object):
         raw_val = raw_val[0]
 
         if isinstance(raw_val, list):
-            return [self.unpack_single_value(data, valtype)
+            return [self.unpack_single_value(data, valtype, valtype_origin)
                     for data in raw_val]
 
-        return self.unpack_single_value(raw_val, valtype)
+        return self.unpack_single_value(raw_val, valtype, valtype_origin)
 
-    def unpack_single_value(self, data, cimtype):
+    def unpack_single_value(self, data, cimtype, cimtype_origin):
         """
         Unpack a single (non-array) CIM typed string value of any CIM type
         except 'reference' and return it as a CIM data type object, or Python
@@ -2160,15 +2171,25 @@ class TupleParser(object):
 
         cimtype (string): CIM data type name (e.g. 'datetime') except
           'reference', or None (in which case a numeric value is assumed).
+
+        cimtype_origin (string): Human readable text explaining the origin of
+          the cimtype argument, for error messages.
         """
         if cimtype in ('string', 'char16', 'datetime'):
-            return self.unpack_string(data, cimtype)
+            return self.unpack_string(data, cimtype, cimtype_origin)
         elif cimtype == 'boolean':
             return self.unpack_boolean(data)
-        # None or 'numeric'
-        return self.unpack_numeric(data, cimtype)
+        elif cimtype is None or NUMERIC_CIMTYPE_PATTERN.match(cimtype):
+            return self.unpack_numeric(data, cimtype, cimtype_origin)
+        else:
+            # Note that 'reference' is not allowed for this function.
+            raise ParseError(
+                _format("Invalid CIM type found: {0!A}; "
+                        "Type origin: {1}",
+                        cimtype, cimtype_origin),
+                conn_id=self.conn_id)
 
-    def unpack_string(self, data, cimtype):
+    def unpack_string(self, data, cimtype, cimtype_origin):
         """
         Unpack a CIM-XML string value of one of the CIM types ('string',
         'char16', 'datetime') and return it as a CIM data type object, or None.
@@ -2177,6 +2198,9 @@ class TupleParser(object):
           None is returned).
 
         cimtype (string): CIM data type name (e.g. 'datetime').
+
+        cimtype_origin (string): Human readable text explaining the origin of
+          the cimtype argument, for error messages.
         """
 
         if data is None:
@@ -2213,8 +2237,9 @@ class TupleParser(object):
                     conn_id=self.conn_id)
         else:
             raise ParseError(
-                _format("Invalid CIM type name {0!A} for string: {1!A}",
-                        cimtype, data),
+                _format("Invalid CIM type name {0!A} for string: {1!A}; "
+                        "Type origin: {2}",
+                        cimtype, data, cimtype_origin),
                 conn_id=self.conn_id)
         return value
 
@@ -2249,7 +2274,7 @@ class TupleParser(object):
                 _format("Invalid boolean value {0!A}", data),
                 conn_id=self.conn_id)
 
-    def unpack_numeric(self, data, cimtype):
+    def unpack_numeric(self, data, cimtype, cimtype_origin):
         """
         Unpack a string value of a numeric CIM type and return its CIM data
         type object, or None.
@@ -2259,6 +2284,9 @@ class TupleParser(object):
 
         cimtype (string): CIM data type name (e.g. 'uint8'), or None (in which
           case the value is returned as a Python int/long or float).
+
+        cimtype_origin (string): Human readable text explaining the origin of
+          the cimtype argument, for error messages.
         """
 
         if data is None:
@@ -2287,7 +2315,8 @@ class TupleParser(object):
                     value = float(data)
                 except ValueError:
                     raise ParseError(
-                        _format("Invalid numeric value {0!A}", data),
+                        _format("Invalid numeric value {0!A}; "
+                                "Type origin: {1}", data, cimtype_origin),
                         conn_id=self.conn_id)
 
         # Convert the Python number into a CIM data type
@@ -2300,7 +2329,8 @@ class TupleParser(object):
             value = CIMType(value)
         except ValueError as exc:
             raise ParseError(
-                str(exc),
+                _format("Cannot convert value {0!A} to numeric CIM type {1}; "
+                        "Type origin: {2}", exc, CIMType, cimtype_origin),
                 conn_id=self.conn_id)
 
         return value

